### PR TITLE
Add click instrumentation

### LINKED ISSUE #2926: flask: create spans for CLI commands
### What problem do you want to solve?

app.cli.command are not instrumented

```python
@app.cli.command()
def create_index():
    """Create or re-create the Elasticsearch index."""
    basedir = os.path.abspath(os.path.dirname(__file__))
    sys.path.append(f"{basedir}/../")

    from data import index_data

    index_data.main()
```

### Describe the solution you'd like

I would like to see an internal (likely root) span with the same name as the function.

### Describe alternatives you've considered

I would like to have an auto-instrumentation alternative to creating a span for an app.cli.command like so:

```python
@app.cli.command()
# Currently, flask auto-instrumentation does not trace CLI commands
@tracer.start_as_current_span("create_index")
def create_index():
    """Create or re-create the Elasticsearch index."""
    basedir = os.path.abspath(os.path.dirname(__file__))
    sys.path.append(f"{basedir}/../")

    from data import index_data

    index_data.main()
```

### Additional Context

_No response_

### Would you like to implement a fix?

No

### PR DESCRIPTION
**Pull request by xrmx**
# Description

This adds a basic instrumentation to trace click commands (as used by flask too) as internal spans.

Fixes #2926

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] tox -e py310-test-instrumentation-click

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+def _command_invoke_wrapper(wrapped, instance, args, kwargs, tracer):
+    # Subclasses of Command include groups and CLI runners, but
+    # we only want to instrument the actual commands which are
+    # instances of Command itself.
+    if instance.__class__ != click.Command:
+        return wrapped(*args, **kwargs)
+
+    ctx = args[0]
+    span_name = ctx.info_name
+    span_attributes = {}
```
Comment (tammy-baylis-swi) [inline instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:70]: Would it be possible to also capture `@click.option` if any exist, maybe as a span attribute?

File: instrumentation/opentelemetry-instrumentation-click/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  "Programming Language :: Python :: 3",
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: 3.12",
+]
+dependencies = [
+  "opentelemetry-api ~= 1.12",
+  "opentelemetry-semantic-conventions == 0.50b0.dev",
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-click/pyproject.toml:29]: Are we missing wrapt as dependency?

File: instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py (hunk 1 above)
Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:70]: Do you think using process attributes is fine?
https://opentelemetry.io/docs/specs/semconv/attributes-registry/process/

Comment (tammy-baylis-swi) [reply instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:70]: Yes I think that will be a good fit. If semconv updates later with more CLI-specific attributes then we can change later.

File: instrumentation/opentelemetry-instrumentation-click/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  "wrapt >= 1.0.0, < 2.0.0",
+]
+
+[project.optional-dependencies]
+instruments = [
+  "click < 9.0.0",
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-click/pyproject.toml:35]: Do we want to have a lower bound of supportability? Maybe > 8.0.0?

File: instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    with tracer.start_as_current_span(
+        name=span_name,
+        kind=trace.SpanKind.INTERNAL,
+        attributes=span_attributes,
+    ) as span:
+        try:
+            return wrapped(*args, **kwargs)
+        except Exception as exc:
+            span.set_status(StatusCode.ERROR, str(exc))
+            span.set_attribute(ERROR_TYPE, exc.__class__.__qualname__)
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:81]: span.is_recording() check would be good here.

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:81]: There may be a chance that adding `span.is_recording()` conditionals for single attributes causes more overhead than just adding the attribute itself :sweat_smile:

File: instrumentation/opentelemetry-instrumentation-click/pyproject.toml (hunk 3 above)
Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-click/pyproject.toml:35]: Good catch, flask requires >= 8.1.3 so I'll go with that too.

File: instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py (hunk 1 above)
Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:70]: Chatted a bit the semconv people, they said the attribute is fine but they asked to create docs formalizing CLI spans. Will take a look at this monday.

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:70]: @tammy-baylis-swi draft PR here https://github.com/open-telemetry/semantic-conventions/pull/1588

File: instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def _command_invoke_wrapper(wrapped, instance, args, kwargs, tracer):
+    # Subclasses of Command include groups and CLI runners, but
+    # we only want to instrument the actual commands which are
+    # instances of Command itself.
+    if instance.__class__ != click.Command:
+        return wrapped(*args, **kwargs)
+
+    ctx = args[0]
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:68]: Curious, what would be the value that is inside `args[0]`? For example in the [click sample](https://click.palletsprojects.com/en/stable/), would the function name `hello` be there or the actual python call `python hello.py`?

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py:68]: ctx is a click Context, its attribute `info_name` would be `hello`

**Review by tammy-baylis-swi — APPROVED** [review]
Thank you for all the semconv work. I think this looks good

File: .github/workflows/test_0.yml [hunk 6]
```diff
@@ -4407,111 +4515,3 @@ jobs:
 
       - name: Run tests
         run: tox -e py312-test-instrumentation-psycopg2 -- -ra
-
-  py38-test-instrumentation-psycopg_ubuntu-latest:
```
[the remark this answers is not in the document]
Comment (xrmx) [reply .github/workflows/test_0.yml:4411]: Yes, it has been moved to test_1.yml

**Review by emdneto — APPROVED** [review]

**Review by lzchen — APPROVED** [review]

**Comment by xrmx** [comment]
Semconv for cli spans PR has been merged.

**Pull request merged**